### ISSUE TITLE
Secondary Sort for Reduce and CoGroup is not working

### ISSUE THREAD

**Issue by AHeise** [stratosphere/stratosphere#206]
Please reenable the secondary ordering as it is essential for data cleansing algorithms.

To reproduce, given input:
(1, x, b)
(1, y, c)
(1, z, a)
And reducer on first field with secondary sort on last, I would expect to get the ordering
(1, z, a)
(1, x, b)
(1, y, c)
But I receive the original order.

I could provide a TestPlan test case.

**Comment by StephanEwen** [comment]
I will take a look at that in a bit.

**Comment by StephanEwen** [comment]
Fixed in 633e3e6a7ed8bc903a749b7f21902c5b4ec5099b

**Comment by AHeise** [comment]
Secondary Sort works only for PactRecord-based programs. 

It boils down to the basic issue that GenericReduceContract do not have Orderings. Hence, only ReduceContracts have a groupOrder that is passed as a GroupProperty by ReduceNode to the optimizer.

The following line in ReduceNode needs to be fixed.

``` java
                // check if we can work with a grouping (simple reducer), or if we need ordering because of a group order
                Ordering groupOrder = null;
                if (getPactContract() instanceof ReduceContract) {
                        groupOrder = ((ReduceContract) getPactContract()).getGroupOrder();
                        if (groupOrder != null && groupOrder.getNumberOfFields() == 0) {
                                groupOrder = null;
                        }
                }
```

I either propose to let GenericReduceContract have group orders or provide an interface for implementing classes that they have such an order. I vote for former, because such Orderings do not assume a tuple model but only assume that the keys can be enumerated and are thus general enough for arbitrary data models.
